src/VisualStudio/Core/Def/Implementation/InheritanceMargin/InheritanceGlyphManager.cs

```diff
@@ -0,0 +1,186 @@
+    /// </summary>
+    internal class InheritanceGlyphManager
+    {
+        private readonly IWpfTextView _textView;
+        private readonly IThreadingContext _threadingContext;
+        private readonly IStreamingFindUsagesPresenter _streamingFindUsagesPresenter;
+        private readonly ClassificationTypeMap _classificationTypeMap;
+        private readonly IClassificationFormatMap _classificationFormatMap;
+        private readonly IUIThreadOperationExecutor _operationExecutor;
+        private readonly IEditorFormatMap _editorFormatMap;
+        private Dictionary<InheritanceMarginGlyph, SnapshotSpan> _glyphToTaggedSpan;
+        private readonly Canvas _glyphsContainer;
+
+        // We want to our glyphs to have the same background color as the glyphs in GlyphMargin
+        private const string GlyphMarginName = "Indicator Margin";
+
+        public InheritanceGlyphManager(
+            IWpfTextView textView,
+            IThreadingContext threadingContext,
+            IStreamingFindUsagesPresenter streamingFindUsagesPresenter,
+            ClassificationTypeMap classificationTypeMap,
+            IClassificationFormatMap classificationFormatMap,
+            IUIThreadOperationExecutor operationExecutor,
+            IEditorFormatMap editorFormatMap,
+            Canvas canvas)
+        {
+            _textView = textView;
+            _threadingContext = threadingContext;
+            _streamingFindUsagesPresenter = streamingFindUsagesPresenter;
+            _classificationTypeMap = classificationTypeMap;
+            _classificationFormatMap = classificationFormatMap;
+            _operationExecutor = operationExecutor;
+            _editorFormatMap = editorFormatMap;
+            _glyphsContainer = canvas;
+            _editorFormatMap.FormatMappingChanged += FormatMappingChanged;
+
+            _glyphToTaggedSpan = new Dictionary<InheritanceMarginGlyph, SnapshotSpan>();
+            _textView.Closed += OnClosed;
+            UpdateBackgroundColor();
+        }
+
+        public void AddGlyph(InheritanceMarginTag tag, SnapshotSpan span)
+        {
+            var lines = _textView.TextViewLines;
+            if (GetStartingLine(lines, span) is IWpfTextViewLine line)
+            {
+                var glyph = CreateNewGlyph(tag);
+                glyph.Height = 17;
+                glyph.Width = 17;
+                SetTop(line, glyph);
+                _glyphToTaggedSpan[glyph] = span;
+                _glyphsContainer.Children.Add(glyph);
+            }
+        }
+
+        public void RemoveGlyph(SnapshotSpan snapshotSpan)
+        {
+            var marginsToRemove = _glyphToTaggedSpan
+                .Where(kvp => snapshotSpan.IntersectsWith(kvp.Value))
+                .ToImmutableArray();
+            foreach (var (margin, span) in marginsToRemove)
+            {
+                _glyphsContainer.Children.Remove(margin);
+                _glyphToTaggedSpan.Remove(margin);
```

Comment:
need threading doucmentation.  otherwise this looks like it could be bad if multiple threads call this.

Reply:
All the code in this class should be called by UI thread only. Will doc these info

Reply:
I have added AssertIsForeground to all these methods.